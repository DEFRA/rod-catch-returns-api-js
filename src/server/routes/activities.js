import {
  Activity,
  Catch,
  River,
  SmallCatch,
  SmallCatchCount,
  Submission
} from '../../entities/index.js'
import {
  createActivitySchema,
  updateActivitySchema
} from '../../schemas/activities.schema.js'
import {
  extractRiverId,
  extractSubmissionId
} from '../../utils/entity-utils.js'
import { handleNotFound, handleServerError } from '../../utils/server-utils.js'
import { StatusCodes } from 'http-status-codes'
import logger from '../../utils/logger-utils.js'
import { mapActivityToResponse } from '../../mappers/activities.mapper.js'
import { mapCatchToResponse } from '../../mappers/catches.mapper.js'
import { mapRiverToResponse } from '../../mappers/river.mapper.js'
import { mapSmallCatchToResponse } from '../../mappers/small-catches.mapper.js'
import { mapSubmissionToResponse } from '../../mappers/submission.mapper.js'
import { sequelize } from '../../services/database.service.js'

const BASE_ACTIVITIES_URL = '/activities/{activityId}'

export default [
  {
    method: 'POST',
    path: '/activities',
    options: {
      /**
       * Create an activity for a submission in the database
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.payload.submission - The submission id prefixed with submissions/
       *     @param {string} request.payload.daysFishedWithMandatoryRelease - The number of days fished during the mandatory release period
       *     @param {string} request.payload.daysFishedOther - The number of days fished at other times during the season
       *     @param {string} request.payload.river - The submission id prefixed with rivers/
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link Activity}
       */
      handler: async (request, h) => {
        try {
          const {
            submission,
            daysFishedWithMandatoryRelease,
            daysFishedOther,
            river
          } = request.payload

          const activityData = {
            submission_id: extractSubmissionId(submission),
            river_id: extractRiverId(river),
            daysFishedWithMandatoryRelease,
            daysFishedOther,
            version: new Date()
          }

          logger.info('Creating activity with details', activityData)

          const createdActivity = await Activity.create(activityData)

          const response = mapActivityToResponse(createdActivity.toJSON())

          return h.response(response).code(StatusCodes.CREATED)
        } catch (error) {
          return handleServerError('Error creating activity', error, h)
        }
      },
      validate: {
        payload: createActivitySchema,
        options: { entity: 'Activity' }
      },
      description: 'Create an activity for a submission in the database',
      notes: 'Create an activity for a submission in the database',
      tags: ['api', 'activities']
    }
  },
  {
    method: 'GET',
    path: `${BASE_ACTIVITIES_URL}/river`,
    options: {
      /**
       * Retrieve the river associated with an activity
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.params.activityId - The activity id
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link River}
       */
      handler: async (request, h) => {
        try {
          const activityId = request.params.activityId
          const activity = await Activity.findOne({
            where: { id: activityId },
            include: [
              {
                model: River,
                required: true
              }
            ]
          })

          if (!activity) {
            return handleNotFound(
              'Activity not found or has no associated river',
              h
            )
          }

          const mappedRiver = mapRiverToResponse(activity.River.toJSON())

          return h.response(mappedRiver).code(StatusCodes.OK)
        } catch (error) {
          return handleServerError(
            'Error fetching river for activity',
            error,
            h
          )
        }
      },
      description:
        'Retrieve the river associated with an activity in the database',
      notes: 'Retrieve river associated with an activity in the database',
      tags: ['api', 'activities']
    }
  },
  {
    method: 'GET',
    path: `${BASE_ACTIVITIES_URL}/smallCatches`,
    options: {
      /**
       * Retrieve the small catches associated with an activity
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.params.activityId - The activity id
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link SmallCatch}
       */
      handler: async (request, h) => {
        try {
          const activityId = request.params.activityId
          const activityWithCatches = await Activity.findOne({
            where: { id: activityId },
            include: [
              {
                model: SmallCatch,
                include: [
                  {
                    association: 'counts'
                  }
                ]
              }
            ]
          })

          if (!activityWithCatches) {
            return handleNotFound(
              `Small catches not found for ${activityId}`,
              h
            )
          }

          const mappedSmallCatches = (
            activityWithCatches.SmallCatches || []
          ).map((smallCatch) => mapSmallCatchToResponse(smallCatch))

          return h
            .response({
              _embedded: {
                smallCatches: mappedSmallCatches
              }
            })
            .code(StatusCodes.OK)
        } catch (error) {
          return handleServerError('Error fetching small catches', error, h)
        }
      },
      description:
        'Retrieve the small catches associated with an activity in the database',
      notes: 'Retrieve small catches with an activity in the database',
      tags: ['api', 'activities']
    }
  },
  {
    method: 'GET',
    path: `${BASE_ACTIVITIES_URL}/catches`,
    options: {
      /**
       * Retrieve the catches (salmon and large sea trout) associated with an activity
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.params.activityId - The activity id
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link Catch}
       */
      handler: async (request, h) => {
        try {
          const activityId = request.params.activityId
          const activityWithCatches = await Activity.findOne({
            where: { id: activityId },
            include: [
              {
                model: Catch
              }
            ]
          })

          if (!activityWithCatches) {
            return handleNotFound(`Catches not found for ${activityId}`, h)
          }

          const mappedCatches = (activityWithCatches.Catches || []).map(
            (catchEntity) => mapCatchToResponse(catchEntity)
          )

          return h
            .response({
              _embedded: {
                catches: mappedCatches
              }
            })
            .code(StatusCodes.OK)
        } catch (error) {
          return handleServerError('Error fetching catches', error, h)
        }
      },
      description:
        'Retrieve the catches (salmon and large sea trout) associated with an activity in the database',
      notes:
        'Retrieve catches (salmon and large sea trout) with an activity in the database',
      tags: ['api', 'activities']
    }
  },
  {
    method: 'GET',
    path: BASE_ACTIVITIES_URL,
    options: {
      /**
       * Retrieve an activity by its ID
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.params.activityId - The activity id
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link Activity}
       */
      handler: async (request, h) => {
        try {
          const activityId = request.params.activityId
          const activity = await Activity.findOne({
            where: { id: activityId }
          })

          if (!activity) {
            return handleNotFound(`Activity not found for ID: ${activityId}`, h)
          }

          const mappedActivity = mapActivityToResponse(activity.toJSON())

          return h.response(mappedActivity).code(StatusCodes.OK)
        } catch (error) {
          return handleServerError('Error fetching activity by ID', error, h)
        }
      },
      description: 'Retrieve an activity by its ID',
      notes: 'Retrieve an activity from the database by its ID',
      tags: ['api', 'activities']
    }
  },
  {
    method: 'DELETE',
    path: BASE_ACTIVITIES_URL,
    options: {
      /**
       * Delete an activity by ID
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.params.activityId - The activity id
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response indicating success or failure
       */
      handler: async (request, h) => {
        const activityId = request.params.activityId

        // Begin transaction for atomic operation
        const transaction = await sequelize.transaction()

        try {
          // Find IDs of all SmallCatch records associated with the Activity
          logger.info(
            'Deleting activity with id:%s and related records',
            activityId
          )
          const smallCatchIds = (
            await SmallCatch.findAll({
              attributes: ['id'],
              where: { activity_id: activityId },
              transaction
            })
          )?.map((smallCatch) => smallCatch.id)

          // Delete associated SmallCatchCount
          await SmallCatchCount.destroy({
            where: { small_catch_id: smallCatchIds },
            transaction
          })

          // Delete associated SmallCatches
          await SmallCatch.destroy({
            where: { activity_id: activityId },
            transaction
          })

          // Delete associated Catches
          await Catch.destroy({
            where: { activity_id: activityId },
            transaction
          })

          // Delete the Activity
          const deletedCount = await Activity.destroy({
            where: { id: activityId },
            transaction
          })

          if (deletedCount === 0) {
            await transaction.rollback()
            return handleNotFound(`Activity with ID ${activityId} not found`, h)
          }

          // Commit transaction
          await transaction.commit()

          logger.info(
            'Deleted activity with id: %s and related records',
            activityId
          )
          return h.response().code(StatusCodes.NO_CONTENT)
        } catch (error) {
          await transaction.rollback()
          return handleServerError('Error deleting activity', error, h)
        }
      },
      description: 'Delete an activity by ID',
      notes: 'Deletes an activity from the database by its ID',
      tags: ['api', 'activities']
    }
  },
  {
    method: 'PATCH',
    path: BASE_ACTIVITIES_URL,
    options: {
      /**
       * Update a activities in the database using the activities ID
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.params.activityId - The ID of the activity to update
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link Activity}
       */
      handler: async (request, h) => {
        const { activityId } = request.params
        const { daysFishedWithMandatoryRelease, daysFishedOther, river } =
          request.payload

        try {
          const activity = await Activity.findByPk(activityId)

          if (!activity) {
            return handleNotFound(`Activity not found for ${activityId}`, h)
          }

          const activityData = {
            river_id: river ? extractRiverId(river) : undefined,
            daysFishedWithMandatoryRelease,
            daysFishedOther,
            version: new Date()
          }

          logger.info(
            `Updating activity ${activityId} with details`,
            activityData
          )

          // if a value is undefined, it is not updated by Sequelize
          const updatedActivity = await activity.update(activityData)

          const mappedActivity = mapActivityToResponse(updatedActivity.toJSON())

          return h.response(mappedActivity).code(StatusCodes.OK)
        } catch (error) {
          return handleServerError('Error updating activity', error, h)
        }
      },
      validate: {
        payload: updateActivitySchema,
        options: { entity: 'Activity' }
      },
      description: 'Update an activity',
      notes: 'Update an activity',
      tags: ['api', 'activities']
    }
  },
  {
    method: 'GET',
    path: `${BASE_ACTIVITIES_URL}/submission`,
    options: {
      /**
       * Retrieve the submission associated with an activity
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.params.activityId - The activity id
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link Submission}
       */
      handler: async (request, h) => {
        try {
          const activityId = request.params.activityId
          const activity = await Activity.findOne({
            where: { id: activityId },
            include: [
              {
                model: Submission,
                required: true
              }
            ]
          })

          if (!activity) {
            return handleNotFound(
              'Activity not found or has no associated submission',
              h
            )
          }

          const mappedSubmission = mapSubmissionToResponse(
            activity.Submission.toJSON()
          )

          return h.response(mappedSubmission).code(StatusCodes.OK)
        } catch (error) {
          return handleServerError(
            'Error fetching submission for activity',
            error,
            h
          )
        }
      },
      description:
        'Retrieve the submission associated with an activity in the database',
      notes: 'Retrieve submission associated with an activity in the database',
      tags: ['api', 'activities']
    }
  }
]
