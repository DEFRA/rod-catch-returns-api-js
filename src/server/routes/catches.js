import {
  mapCatchToResponse,
  mapRequestToCatch
} from '../../mappers/catches.mapper.js'
import { Catch } from '../../entities/index.js'
import { StatusCodes } from 'http-status-codes'
import logger from '../../utils/logger-utils.js'

export default [
  {
    method: 'POST',
    path: '/catches',
    options: {
      /**
       * Create a catch (salmon and large sea trout) for an activity in the database
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.payload.activity - The activity id prefixed with activities/
       *     @param {string} request.payload.dateCaught - The ISO 8601 date the catch was recorded (e.g., "2024-06-24T00:00:00+01:00").
       *     @param {string} request.payload.species - The species ID prefixed with "species/".
       *     @param {Object} request.payload.mass - The mass of the catch.
       *     @param {number} request.payload.mass.kg - The mass in kilograms.
       *     @param {number} request.payload.mass.oz - The mass in ounces.
       *     @param {string} request.payload.mass.type - The mass type, typically "IMPERIAL" or "METRIC".
       *     @param {string} request.payload.method - The method ID prefixed with "methods/".
       *     @param {boolean} request.payload.released - Indicates if the catch was released.
       *     @param {boolean} request.payload.onlyMonthRecorded - Indicates if only the month was recorded, to allow FMT users to report on the default dates
       *     @param {boolean} request.payload.noDateRecorded - Indicates if no specific date was recorded, to allow FMT users to report on the default month
       *     @param {string} request.payload.reportingExclude - Is this entry excluded from reporting
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link Activity}
       */
      handler: async (request, h) => {
        try {
          const catchData = mapRequestToCatch(request.payload)

          const createdCatch = await Catch.create(catchData)

          const catchResponse = mapCatchToResponse(
            request,
            createdCatch.toJSON()
          )

          return h.response(catchResponse).code(StatusCodes.CREATED)
        } catch (error) {
          logger.error('Error create catch:', error)
          return h
            .response({ error: 'Unable to create catch' })
            .code(StatusCodes.INTERNAL_SERVER_ERROR)
        }
      },
      description:
        'Create a catch (salmon and large sea trout) for an activity in the database',
      notes:
        'Create a catch (salmon and large sea trout) for an activity in the database',
      tags: ['api', 'catches']
    }
  }
]
