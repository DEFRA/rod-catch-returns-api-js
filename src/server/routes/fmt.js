import { authorize, signIn } from '../../services/oidc-service.js'

export default [
  {
    path: '/oidc/authorize',
    method: 'GET',
    handler: authorize,
    options: { auth: false }
  },
  {
    path: '/oidc/callback',
    method: 'POST',
    handler: signIn,
    options: { auth: false }
  }
]
