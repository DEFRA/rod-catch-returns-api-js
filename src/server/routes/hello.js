export default [
  {
    method: 'GET',
    path: '/hello',
    /**
     * Handler that return hello.
     *
     * @param {import('@hapi/hapi').Request request - The Hapi request object
     * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
     * @returns {import('@hapi/hapi').ResponseObject} - A response containing hello
     */
    handler: async (request, h) => {
      return h.response('Hello mate').code(200)
    }
  }
]
