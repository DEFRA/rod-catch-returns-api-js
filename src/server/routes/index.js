import Catchments from './catchments.js'
import Licences from './licences.js'
import Region from './regions.js'
import Rivers from './rivers.js'
import Species from './species.js'

import Static from './static.js'

export const apiPrefixRoutes = [].concat(Licences, Region, Rivers, Catchments, Species)

export const rootRoutes = [].concat(Static)
