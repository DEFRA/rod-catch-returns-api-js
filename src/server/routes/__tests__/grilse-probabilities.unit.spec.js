import {
  deleteGrilseProbabilitiesForSeasonAndGate,
  generateCsvFromGrilseProbabilities,
  getGrilseProbabilitiesBySeasonRange,
  isGrilseProbabilityExistsForSeasonAndGate,
  processGrilseProbabilities,
  validateAndParseCsvFile
} from '../../../services/grilse-probabilities.service.js'
import {
  getMockResponseToolkit,
  getServerDetails
} from '../../../test-utils/server-test-utils.js'
import {
  handleNotFound,
  handleServerError
} from '../../../utils/server-utils.js'
import { GrilseProbability } from '../../../entities/index.js'
import { GrilseValidationError } from '../../../models/grilse-validation-error.model.js'

import routes from '../grilse-probabilities.js'

jest.mock('../../../services/grilse-probabilities.service.js')
jest.mock('../../../entities/index.js')
jest.mock('../../../utils/logger-utils.js')
jest.mock('../../../utils/server-utils.js')

const [
  {
    options: { handler: getAllGrilseProbabilitiesHandler }
  },
  {
    options: { handler: deleteGrilseProbabilityByIdHandler }
  },
  {
    options: { handler: uploadGrilseProbabilitiesHandler }
  },
  {
    options: { handler: getGrilseProbabilitiesHandler }
  }
] = routes

const NOT_FOUND_SYMBOL = Symbol('NOT_FOUND')
const SERVER_ERROR_SYMBOL = Symbol('SERVER_ERROR')

handleNotFound.mockReturnValue(NOT_FOUND_SYMBOL)
handleServerError.mockReturnValue(SERVER_ERROR_SYMBOL)

describe('grilse-probabilities.unit', () => {
  describe('GET /grilseProbabilities', () => {
    const getGrilseProbabilitiesData = () => [
      {
        id: '2',
        season: 2025,
        month: 4,
        massInPounds: 6,
        probability: '1.0000000000000000',
        createdAt: '2025-09-01T08:57:34.479Z',
        updatedAt: '2025-09-01T08:57:34.479Z',
        version: '2025-09-01T08:57:34.479Z',
        gate_id: '2'
      },
      {
        id: '2',
        season: 2025,
        month: 5,
        massInPounds: 6,
        probability: '0.9920000000000000',
        createdAt: '2025-09-01T08:57:34.479Z',
        updatedAt: '2025-09-01T08:57:34.479Z',
        version: '2025-09-01T08:57:34.479Z',
        gate_id: '2'
      }
    ]

    afterEach(() => {
      jest.clearAllMocks()
    })

    it('should return a 200 status code and the species if the call to fetch all species is successful', async () => {
      GrilseProbability.findAll.mockResolvedValueOnce(
        getGrilseProbabilitiesData()
      )

      const result = await getAllGrilseProbabilitiesHandler(
        getServerDetails(),
        getMockResponseToolkit()
      )

      expect(result.payload).toMatchSnapshot()
      expect(result.statusCode).toBe(200)
    })

    it('should call handleServerError if an error occurs while fetching species', async () => {
      const error = new Error('Database error')
      GrilseProbability.findAll.mockRejectedValueOnce(error)
      const h = getMockResponseToolkit()

      await getAllGrilseProbabilitiesHandler(getServerDetails(), h)

      expect(handleServerError).toHaveBeenCalledWith(
        'Error fetching grilse probabilities',
        error,
        h
      )
    })

    it('should return an error response if an error occurs while fetching species', async () => {
      const error = new Error('Database error')
      GrilseProbability.findAll.mockRejectedValueOnce(error)

      const result = await getAllGrilseProbabilitiesHandler(
        getServerDetails(),
        getMockResponseToolkit()
      )

      expect(result).toBe(SERVER_ERROR_SYMBOL)
    })
  })

  describe('DELETE /grilseProbabilities/{grilseProbabilityId}', () => {
    const getDeleteRequest = (grilseProbabilityId) =>
      getServerDetails({ params: { grilseProbabilityId } })

    it('should return a 204 status code on successful deletion', async () => {
      GrilseProbability.destroy.mockResolvedValueOnce(1)

      const result = await deleteGrilseProbabilityByIdHandler(
        getDeleteRequest('3'),
        getMockResponseToolkit()
      )

      expect(result.statusCode).toBe(204)
    })

    it('should return an empty response on successful deletion', async () => {
      GrilseProbability.destroy.mockResolvedValueOnce(1)

      const result = await deleteGrilseProbabilityByIdHandler(
        getDeleteRequest('3'),
        getMockResponseToolkit()
      )

      expect(result.payload).toBeUndefined()
    })

    it('should call GrilseProbability.destroy with the correct parameters', async () => {
      const grilseProbabilityId = '3'
      GrilseProbability.destroy.mockResolvedValueOnce(1)

      await deleteGrilseProbabilityByIdHandler(
        getDeleteRequest(grilseProbabilityId),
        getMockResponseToolkit()
      )

      expect(GrilseProbability.destroy).toHaveBeenCalledWith({
        where: { id: grilseProbabilityId }
      })
    })

    it('should return a 404 status code if the GrilseProbability does not exist', async () => {
      GrilseProbability.destroy.mockResolvedValueOnce(0)

      const result = await deleteGrilseProbabilityByIdHandler(
        getDeleteRequest('3'),
        getMockResponseToolkit()
      )

      expect(result.statusCode).toBe(404)
    })

    it('should call handleServerError if an error occurs during GrilseProbability.destroy', async () => {
      const error = new Error('Database error')

      GrilseProbability.destroy.mockRejectedValueOnce(error)
      const h = getMockResponseToolkit()

      await deleteGrilseProbabilityByIdHandler(getDeleteRequest('3'), h)

      expect(handleServerError).toHaveBeenCalledWith(
        'Error deleting grilse probability',
        error,
        h
      )
    })
  })

  describe('POST /reporting/reference/grilse-probabilities/{season}/{gate}', () => {
    const mockCsvData = 'Weight,January,February\n10,0.2,0.3\n15,0.5,0.6'
    const getMockRequest = ({
      season = '2024',
      gate = '1',
      overwrite = false
    } = {}) => ({
      params: { season, gate },
      query: { overwrite },
      payload: Buffer.from(mockCsvData)
    })

    beforeEach(() => {
      jest.clearAllMocks()
    })

    it('should return 201 amd create the records when the CSV is processed successfully', async () => {
      isGrilseProbabilityExistsForSeasonAndGate.mockResolvedValueOnce(false)
      validateAndParseCsvFile.mockResolvedValueOnce([
        ['Weight', 'June'],
        ['1', '1.0'],
        ['2', '1.0']
      ])
      processGrilseProbabilities.mockReturnValueOnce([
        {
          season: '2024',
          gate_id: '1',
          massInPounds: 10,
          probability: 0.2,
          version: new Date()
        }
      ])
      GrilseProbability.bulkCreate.mockResolvedValueOnce()

      const result = await uploadGrilseProbabilitiesHandler(
        getServerDetails(getMockRequest()),
        getMockResponseToolkit()
      )

      expect(result.statusCode).toBe(201)
      expect(GrilseProbability.bulkCreate).toHaveBeenCalled()
    })

    it('should return 409 and an error message if existing data is found and overwrite is not set', async () => {
      isGrilseProbabilityExistsForSeasonAndGate.mockResolvedValueOnce(true)

      const result = await uploadGrilseProbabilitiesHandler(
        getServerDetails(getMockRequest({ overwrite: false })),
        getMockResponseToolkit()
      )

      expect(result.statusCode).toBe(409)
      expect(result.payload).toStrictEqual({
        message:
          'Existing data found for the given season and gate but overwrite parameter not set'
      })
    })

    it('should delete existing data and create new if overwrite is set to true', async () => {
      isGrilseProbabilityExistsForSeasonAndGate.mockResolvedValueOnce(true)

      validateAndParseCsvFile.mockResolvedValueOnce([
        ['Weight', 'June'],
        ['1', '1.0'],
        ['2', '1.0']
      ])
      processGrilseProbabilities.mockReturnValueOnce([
        {
          season: '2024',
          gate_id: '1',
          massInPounds: 10,
          probability: 0.2,
          version: new Date()
        }
      ])
      GrilseProbability.bulkCreate.mockResolvedValueOnce()

      await uploadGrilseProbabilitiesHandler(
        getServerDetails(getMockRequest({ overwrite: true })),
        getMockResponseToolkit()
      )

      expect(deleteGrilseProbabilitiesForSeasonAndGate).toHaveBeenCalledWith(
        '2024',
        '1'
      )
      expect(GrilseProbability.bulkCreate).toHaveBeenCalled()
    })

    it('should return a 201 and not create any records if there are no valid probabilities', async () => {
      isGrilseProbabilityExistsForSeasonAndGate.mockResolvedValueOnce(false)
      validateAndParseCsvFile.mockResolvedValueOnce([
        ['Weight', 'June'],
        ['1', '1.0'],
        ['2', '1.0']
      ])
      processGrilseProbabilities.mockReturnValueOnce([])

      const result = await uploadGrilseProbabilitiesHandler(
        getServerDetails(getMockRequest()),
        getMockResponseToolkit()
      )

      expect(result.statusCode).toBe(201)
      expect(GrilseProbability.bulkCreate).not.toHaveBeenCalled()
    })

    it('should handle errors gracefully and return server error response', async () => {
      const error = new Error('Database error')
      isGrilseProbabilityExistsForSeasonAndGate.mockRejectedValueOnce(error)

      const result = await uploadGrilseProbabilitiesHandler(
        getServerDetails(getMockRequest()),
        getMockResponseToolkit()
      )

      expect(handleServerError).toHaveBeenCalledWith(
        'Error uploading grilse probabilities file',
        error,
        expect.anything()
      )
      expect(result).toBe(SERVER_ERROR_SYMBOL)
    })

    it('should return a 400 error if validation fails', async () => {
      validateAndParseCsvFile.mockImplementation(() => {
        throw new GrilseValidationError({ status: 400 })
      })
      const request = {
        params: { season: '2023', gate: '1' },
        query: { overwrite: 'true' },
        payload: { invalid: 'object' } // Object payload
      }

      const result = await uploadGrilseProbabilitiesHandler(
        getServerDetails(request),
        getMockResponseToolkit()
      )
      expect(result.statusCode).toBe(400)
    })

    it('should return an error response with timestamp, message and path if validation fails', async () => {
      validateAndParseCsvFile.mockImplementation(() => {
        throw new GrilseValidationError()
      })
      const request = {
        params: { season: '2023', gate: '1' },
        query: { overwrite: 'true' },
        payload: { invalid: 'object' } // Object payload
      }

      const result = await uploadGrilseProbabilitiesHandler(
        getServerDetails(request),
        getMockResponseToolkit()
      )
      expect(result.payload).toStrictEqual({
        timestamp: expect.any(String),
        status: 400,
        message: 'Validation error',
        path: undefined
      })
    })

    it('should return an error field, if it is returned by GrilseValidationError', async () => {
      validateAndParseCsvFile.mockImplementation(() => {
        throw new GrilseValidationError({
          status: 400,
          error: 'This is an error'
        })
      })
      const request = {
        params: { season: '2023', gate: '1' },
        query: { overwrite: 'true' },
        payload: { invalid: 'object' } // Object payload
      }

      const result = await uploadGrilseProbabilitiesHandler(
        getServerDetails(request),
        getMockResponseToolkit()
      )
      expect(result.payload).toStrictEqual(
        expect.objectContaining({
          error: 'This is an error'
        })
      )
    })

    it('should return an errors array, if it is returned by GrilseValidationError', async () => {
      validateAndParseCsvFile.mockImplementation(() => {
        throw new GrilseValidationError({
          status: 400,
          errors: ['this is an error']
        })
      })
      const request = {
        params: { season: '2023', gate: '1' },
        query: { overwrite: 'true' },
        payload: { invalid: 'object' } // Object payload
      }

      const result = await uploadGrilseProbabilitiesHandler(
        getServerDetails(request),
        getMockResponseToolkit()
      )
      expect(result.payload).toStrictEqual(
        expect.objectContaining({
          errors: ['this is an error']
        })
      )
    })
  })

  describe('GET /reporting/reference/grilse-probabilities/{season}', () => {
    const getMockRequest = (season = '2024') => ({
      params: { season }
    })

    const getMockTypeHeaderResponseToolkit = () => {
      const headers = {}

      const code = (statusCode, payload) => {
        return { statusCode, headers, payload }
      }

      const header = (name, value, payload) => {
        headers[name.toLowerCase()] = value
        return { code: (statusCode) => code(statusCode, payload) }
      }

      return {
        response: jest.fn().mockImplementation((payload) => {
          return {
            type: (contentType) => {
              headers['content-type'] = contentType
              return { header: (name, value) => header(name, value, payload) }
            },
            header: (name, value) => {
              headers[name.toLowerCase()] = value
              return { code: (statusCode) => code(statusCode, payload) }
            },
            code: (statusCode) => {
              return { statusCode, headers, payload }
            }
          }
        })
      }
    }

    const getMockCsvData = () => 'Weight,January\n10,0.2\n15,0.5'

    const getMockGrilseProbabilities = () => [
      { season: 2024, gate_id: '1', massInPounds: 10, probability: 0.2 },
      { season: 2024, gate_id: '1', massInPounds: 15, probability: 0.5 }
    ]

    beforeEach(() => {
      jest.clearAllMocks()
    })

    it('should return 200 and a CSV file when records are found', async () => {
      const mockCsvData = getMockCsvData()
      getGrilseProbabilitiesBySeasonRange.mockResolvedValueOnce(
        getMockGrilseProbabilities()
      )
      generateCsvFromGrilseProbabilities.mockReturnValueOnce(mockCsvData)

      const result = await getGrilseProbabilitiesHandler(
        getServerDetails(getMockRequest()),
        getMockTypeHeaderResponseToolkit()
      )

      expect(result.statusCode).toBe(200)
      expect(result.payload).toBe(mockCsvData)
      expect(result.headers['content-type']).toBe('text/csv')
      expect(result.headers['content-disposition']).toBe(
        'attachment; filename="grilse-probabilities-2024.csv"'
      )
    })

    it('should return the correct headers when records are found', async () => {
      const mockCsvData = getMockCsvData()
      getGrilseProbabilitiesBySeasonRange.mockResolvedValueOnce(
        getMockGrilseProbabilities()
      )
      generateCsvFromGrilseProbabilities.mockReturnValueOnce(mockCsvData)

      const result = await getGrilseProbabilitiesHandler(
        getServerDetails(getMockRequest()),
        getMockTypeHeaderResponseToolkit()
      )

      expect(result.headers['content-type']).toBe('text/csv')
      expect(result.headers['content-disposition']).toBe(
        'attachment; filename="grilse-probabilities-2024.csv"'
      )
    })

    it('should return 404 when no grilse probabilities are found', async () => {
      getGrilseProbabilitiesBySeasonRange.mockResolvedValueOnce([])
      const h = getMockTypeHeaderResponseToolkit()

      const result = await getGrilseProbabilitiesHandler(
        getServerDetails(getMockRequest()),
        h
      )

      expect(handleNotFound).toHaveBeenCalledWith(
        'Grilse probabilities not found for 2024',
        h
      )
      expect(result).toBe(NOT_FOUND_SYMBOL)
    })

    it('should correctly handle a single season', async () => {
      getGrilseProbabilitiesBySeasonRange.mockResolvedValueOnce(
        getMockGrilseProbabilities()
      )
      generateCsvFromGrilseProbabilities.mockReturnValueOnce(getMockCsvData())

      const result = await getGrilseProbabilitiesHandler(
        getServerDetails(getMockRequest('2024')),
        getMockTypeHeaderResponseToolkit()
      )

      expect(result.statusCode).toBe(200)
      expect(getGrilseProbabilitiesBySeasonRange).toHaveBeenCalledWith(
        2024,
        2024
      )
    })

    it('should correctly handle a season range', async () => {
      getGrilseProbabilitiesBySeasonRange.mockResolvedValueOnce(
        getMockGrilseProbabilities()
      )
      generateCsvFromGrilseProbabilities.mockReturnValueOnce(getMockCsvData())

      const result = await getGrilseProbabilitiesHandler(
        getServerDetails(getMockRequest('2023-2024')),
        getMockTypeHeaderResponseToolkit()
      )

      expect(result.statusCode).toBe(200)
      expect(getGrilseProbabilitiesBySeasonRange).toHaveBeenCalledWith(
        2023,
        2024
      )
    })

    it('should handle errors gracefully and return a server error response', async () => {
      const error = new Error('Database error')
      getGrilseProbabilitiesBySeasonRange.mockRejectedValueOnce(error)

      const result = await getGrilseProbabilitiesHandler(
        getServerDetails(getMockRequest()),
        getMockTypeHeaderResponseToolkit()
      )

      expect(handleServerError).toHaveBeenCalledWith(
        'Error retrieving grilse probabilities file',
        error,
        expect.anything()
      )
      expect(result).toBe(SERVER_ERROR_SYMBOL)
    })
  })
})
