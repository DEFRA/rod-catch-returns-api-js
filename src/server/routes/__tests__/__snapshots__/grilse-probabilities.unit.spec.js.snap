// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`grilse-probabilities.unit GET /grilseProbabilities should return a 200 status code and the species if the call to fetch all species is successful 1`] = `
{
  "_embedded": {
    "grilseProbabilities": [
      {
        "_links": {
          "gate": {
            "href": "http://localhost:5000/api/grilseProbabilities/2/gate",
          },
          "grilseProbability": {
            "href": "http://localhost:5000/api/grilseProbabilities/2",
          },
          "self": {
            "href": "http://localhost:5000/api/grilseProbabilities/2",
          },
        },
        "createdAt": "2025-09-01T08:57:34.479Z",
        "id": "2",
        "massInPounds": 6,
        "month": 4,
        "probability": "1.0000000000000000",
        "season": 2025,
        "updatedAt": "2025-09-01T08:57:34.479Z",
      },
      {
        "_links": {
          "gate": {
            "href": "http://localhost:5000/api/grilseProbabilities/2/gate",
          },
          "grilseProbability": {
            "href": "http://localhost:5000/api/grilseProbabilities/2",
          },
          "self": {
            "href": "http://localhost:5000/api/grilseProbabilities/2",
          },
        },
        "createdAt": "2025-09-01T08:57:34.479Z",
        "id": "2",
        "massInPounds": 6,
        "month": 5,
        "probability": "0.9920000000000000",
        "season": 2025,
        "updatedAt": "2025-09-01T08:57:34.479Z",
      },
    ],
  },
}
`;
